lib/core.ts

```diff
@@ -0,0 +1,165 @@
+  verbose: boolean;
+};
+
+const defaultConfig: Partial<Config> = {
+  verifyContracts: false,
+  saveNetworkConfig: false,
+  verbose: false,
+};
+
+export abstract class NeokingdomDAO {
+  config: Config;
+
+  constructor(config: Config) {
+    this.config = config;
+  }
+
+  static _mergeWithDefaultConfig(config: Partial<Config> = {}) {
+    return {
+      ...defaultConfig,
+      ...config,
+    } as Config;
+  }
+
+  public getNextStepFilename() {
+    return `./deployments/${this.config.chainId}.nextstep`;
+  }
+
+  public async setNextStep(n: number) {
+    await writeFile(this.getNextStepFilename(), n.toString());
+  }
+
+  public async getNextStep() {
+    try {
+      return parseInt(await readFile(this.getNextStepFilename(), "utf8"));
+    } catch (e) {
+      if ((e as any).code !== "ENOENT") {
+        throw e;
+      }
+    }
+    return 0;
+  }
+
+  async run<T extends Context>(
+    c: ContextGenerator<T>,
+    s: Sequence<T>,
+    config: { force?: boolean; restart?: boolean } = {}
+  ) {
+    const { force, restart } = config;
+    const sequence = await this._preprocessSequence(c, s);
+    const nextStep = restart ? 0 : await this.getNextStep();
+    await this._executeSequence(c, sequence, nextStep, force);
+  }
+
+  abstract loadContractsPartial(): Promise<Partial<NeokingdomContracts>>;
+
+  async loadContracts(): Promise<NeokingdomContracts> {
+    const contracts = await this.loadContractsPartial();
+    if (isNeokingdomContracts(contracts)) {
+      return contracts;
+    }
+    throw new Error("Missing contracts");
+  }
+
+  abstract deploy(contractName: ContractNames, args?: any[]): Promise<Contract>;
+
+  abstract deployProxy(
+    contractName: ContractNames,
+    args?: any[]
+  ): Promise<Contract>;
+
+  private async _preprocessSequence<T extends Context>(
+    c: ContextGenerator<T>,
+    s: Sequence<T>
+  ) {
+    let sequence: ProcessedSequence<T> = [];
+    for (let i = 0; i < s.length; i++) {
+      // FIXME: Don't know why Awaited<T> is not the same as T
+      const context = (await c(this)) as T;
+      const step = s[i];
+      const expanded = preprocessStep(context, step);
+      sequence = [...sequence, ...expanded];
+    }
+    return sequence;
+  }
+
+  private async _executeSequence<T extends Context>(
+    c: ContextGenerator<T>,
+    s: ProcessedSequence<T>,
+    nextStep = 0,
+    force = false
+  ) {
+    for (let i = nextStep; i < s.length; i++) {
+      const context = await c(this);
+      const step = s[i];
+      if (this.config.verbose) {
+        console.log(`${i + 1}/${s.length}: ${step.toString()}`);
+      }
+      let tx: Contract | ContractTransaction | null = null;
+      try {
+        tx = await step(context);
+      } catch (e) {
+        if (force) {
+          console.error(e);
+        } else {
+          throw e;
+        }
+      }
+      if (this.config.verbose) {
+        console.log();
+      }
+      // FIXME: wait should always be a valid attribute, but it's not
```

Comment:
why so? if `tx` is null, it makes sense that `wait` is not there, right?

Reply:
The issue here is that the `step` function was returning the wrong type. I fixed it in my last commit https://github.com/NeokingdomDAO/contracts/pull/18/commits/41210e9bcc0d42046c0dd6722fb3758c37b58bfb